fdbclient/MasterProxyInterface.cpp

```diff
@@ -0,0 +1,337 @@
+			if (count + mutationSize[i] >= maximumBytes) {
+				++proxy;
+				if (proxy == NUM_PROXIES) {
+					return true;
+				}
+				count = 0;
+			}
+			count += mutationSize[i];
+		}
+		return false;
+	}
+
+public:
+	MutationSequenceDistributor(const int numProxies) : NUM_PROXIES(numProxies) {}
+
+	void insert(int mutationBytes) {
+		mutationSize.push_back(mutationBytes);
+		++numMutations;
+		mutationMaxBytes = std::max(mutationMaxBytes, mutationBytes);
+		totalBytes += mutationBytes;
+	}
+
+	std::vector<int> evaluate() {
+		ASSERT(numMutations > 0);
+
+		std::vector<int> result;
+		result.reserve(numMutations);
+
+		// If there are more number of proxies than number of mutations, distribute N mutations to first N proxies and
+		// leave the remaining proxies empty.
+		if (NUM_PROXIES >= numMutations) {
+			for (int i = 0; i < NUM_PROXIES; ++i) {
+				if (i < numMutations)
+					result.push_back(i);
+				else
+					result.push_back(numMutations);
+			}
+			return result;
+		}
+
+		// Try to minimize the number of maximum bytes per proxy
+		boost::integer_range range(mutationMaxBytes, totalBytes);
+		auto func = std::bind(&MutationSequenceDistributor::couldNotDistributeOverProxies, this, std::placeholders::_1,
+		                      std::placeholders::_2);
+		const int MAX_BYTES = *boost::lower_bound(range, 0, func);
+
+		int count = 0;
+		int i = 0;
+
+		// The subversion starts with 1 rather than 0
+		result.push_back(1);
+		for (i = 0; i < numMutations; ++i) {
+			if (count + mutationSize[i] >= MAX_BYTES) {
+				count = 0;
+				// Subversion starts from 1
+				result.push_back(i + 1);
+
+				// remaining number of proxies, note the current one is a proxy has not assigned any bytes of data
+				const int NUM_REMAINING_PROXIES = NUM_PROXIES - result.size() + 1;
+				// remaining mutations has not assigned to the proxies, note the current one is not assigned yet
+				const int NUM_REMAINING_MUTATIONS = numMutations - i;
+				if (NUM_REMAINING_MUTATIONS == NUM_REMAINING_PROXIES) {
+					// evenly distribute the remaining items over the remaining proxies, rather than greedly
+					while (result.size() < static_cast<size_t>(NUM_PROXIES)) {
+						result.push_back(result.back() + 1);
+						break;
+					}
+				}
+			}
+			count += mutationSize[i];
+		}
+
+		return result;
+	}
+};
+
+void distributeMutationsInSequence(const CommitTransactionRequest& request,
+                                   std::vector<CommitTransactionRequest>& splitCommitTxnRequests,
+                                   std::vector<int>& requestMutationStartSubversion) {
+	ASSERT(splitCommitTxnRequests.size() > 0);
+
+	const int NUM_PROXIES = splitCommitTxnRequests.size();
+	const int NUM_MUTATIONS = request.transaction.mutations.size();
+
+	const auto& mutations = request.transaction.mutations;
+
+	MutationSequenceDistributor distributor(NUM_PROXIES);
+	for (int i = 0; i < mutations.size(); ++i) {
+		distributor.insert(mutations[i].param2.size());
```

Comment:
I suppose you should count the whole mutation size, i.e., `mutations[i].expectedSize()`.